File: native_handoff/android/Components.kt
// Components.kt — Just Marriage Jetpack Compose component recipes.
// Mirror the web design system's contracts (see ../components.md). Depend only on
// Color.kt / Dimens.kt / Shape.kt / Type.kt.
package com.justmarriage.design

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// ---------------------------------------------------------------- Button

enum class JMButtonVariant { Primary, Secondary, Ink, Outline, Ghost }
enum class JMButtonSize { Sm, Md, Lg }

@Composable
fun JMButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: JMButtonVariant = JMButtonVariant.Primary,
    size: JMButtonSize = JMButtonSize.Md,
    pill: Boolean = false,
    fullWidth: Boolean = false,
    enabled: Boolean = true,
) {
    val height = when (size) { JMButtonSize.Sm -> 36.dp; JMButtonSize.Md -> 46.dp; JMButtonSize.Lg -> 56.dp }
    val fontSize = when (size) { JMButtonSize.Sm -> 13.sp; JMButtonSize.Md -> 15.sp; JMButtonSize.Lg -> 16.sp }
    val bg = when (variant) {
        JMButtonVariant.Primary -> JMColors.primary
        JMButtonVariant.Secondary -> JMColors.secondary
        JMButtonVariant.Ink -> JMColors.ink
        else -> Color.Transparent
    }
    val fg = when (variant) {
        JMButtonVariant.Primary, JMButtonVariant.Ink -> JMColors.onPrimary
        JMButtonVariant.Secondary -> JMColors.onSecondary
        else -> JMColors.ink
    }
    val shape = if (pill) JMShapes.pill else JMShapes.md
    Box(
        modifier
            .then(if (fullWidth) Modifier.fillMaxWidth() else Modifier)
            .height(height)
            .clip(shape)
            .background(if (enabled) bg else bg.copy(alpha = 0.45f))
            .then(if (variant == JMButtonVariant.Outline) Modifier.border(JMBorder.widthBold, JMColors.ink, shape) else Modifier)
            .clickableNoRipple(enabled, onClick)
            .padding(horizontal = if (size == JMButtonSize.Lg) 30.dp else 22.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(text, color = fg, fontFamily = JMFontFamily.Sans, fontWeight = FontWeight.Bold, fontSize = fontSize)
    }
}

// ---------------------------------------------------------------- Badge

enum class JMBadgeTone { Ink, Pink, Cyan, Yellow, Success, Warning, Error }

@Composable
fun JMBadge(
    text: String,
    tone: JMBadgeTone = JMBadgeTone.Ink,
    soft: Boolean = false,
    tilt: Boolean = false,
    uppercase: Boolean = false,
    modifier: Modifier = Modifier,
) {
    val (bg, fg) = when (tone) {
        JMBadgeTone.Ink     -> if (soft) JMPalette.Ink100 to JMPalette.Ink900   else JMPalette.Ink900 to JMPalette.White
        JMBadgeTone.Pink    -> if (soft) JMPalette.Pink50 to JMPalette.Pink700  else JMPalette.Pink500 to JMPalette.White
        JMBadgeTone.Cyan    -> if (soft) JMPalette.Cyan50 to JMPalette.Cyan700  else JMPalette.CyanBright to JMPalette.Ink900
        JMBadgeTone.Yellow  -> JMPalette.Yellow400 to JMPalette.Ink900
        JMBadgeTone.Success -> if (soft) JMPalette.Green50 to JMPalette.Green500 else JMPalette.Green500 to JMPalette.White
        JMBadgeTone.Warning -> if (soft) JMPalette.Amber50 to JMPalette.Amber500 else JMPalette.Amber500 to JMPalette.Ink900
        JMBadgeTone.Error   -> if (soft) JMPalette.Red50 to JMPalette.Red500     else JMPalette.Red500 to JMPalette.White
    }
    Box(
        modifier
            .rotate(if (tilt) -4f else 0f)
            .clip(JMShapes.xs)
            .background(bg)
            .padding(horizontal = 11.dp, vertical = 4.dp)
    ) {
        Text(
            if (uppercase) text.uppercase() else text,
            color = fg, fontFamily = JMFontFamily.Sans, fontWeight = FontWeight.ExtraBold, fontSize = 13.sp,
        )
    }
}

// ---------------------------------------------------------------- Card

enum class JMCardVariant { Plain, Hard, Tinted }

@Composable
fun JMCard(
    modifier: Modifier = Modifier,
    variant: JMCardVariant = JMCardVariant.Plain,
    tint: Color = JMPalette.Cyan50,
    padding: Dp = JMSpace.x5,
    content: @Composable ColumnScope.() -> Unit,
) {
    val shape = JMShapes.lg
    Box(
        modifier.then(
            if (variant == JMCardVariant.Hard)
                Modifier.drawBehind {
                    val o = 4.dp.toPx()
                    drawRoundRect(JMColors.ink, topLeft = Offset(o, o),
                        size = size, cornerRadius = androidx.compose.ui.geometry.CornerRadius(20.dp.toPx()))
                }
            else Modifier
        )
    ) {
        Column(
            Modifier
                .clip(shape)
                .background(if (variant == JMCardVariant.Tinted) tint else JMColors.surfaceCard)
                .then(
                    when (variant) {
                        JMCardVariant.Hard -> Modifier.border(JMBorder.widthBold, JMColors.borderStrong, shape)
                        JMCardVariant.Plain -> Modifier.border(JMBorder.width, JMColors.borderSubtle, shape)
                        else -> Modifier
                    }
                )
                .padding(padding),
            content = content,
        )
    }
}

// ---------------------------------------------------------------- ProgressRing (match dial)

@Composable
fun JMProgressRing(
    value: Float,              // 0..100
    modifier: Modifier = Modifier,
    size: Dp = 120.dp,
    thickness: Dp = 12.dp,
    sublabel: String? = null,
) {
    val sweep by animateFloatAsState(value.coerceIn(0f, 100f) / 100f * 360f, JMMotion.easeOut(), label = "ring")
    Box(modifier.size(size), contentAlignment = Alignment.Center) {
        androidx.compose.foundation.Canvas(Modifier.size(size)) {
            val stroke = Stroke(width = thickness.toPx(), cap = StrokeCap.Round)
            drawArc(JMPalette.Ink100, 0f, 360f, false, style = Stroke(width = thickness.toPx()))
            drawArc(
                Brush.sweepGradient(listOf(JMColors.primary, JMColors.secondary)),
                -90f, sweep, false, style = stroke,
            )
        }
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text("${value.toInt()}%", fontFamily = JMFontFamily.Display, fontSize = (size.value * 0.3).sp, color = JMColors.ink)
            sublabel?.let {
                Text(it.uppercase(), fontFamily = JMFontFamily.Sans, fontWeight = FontWeight.Bold,
                    fontSize = (size.value * 0.1).sp, color = JMColors.textTertiary)
            }
        }
    }
}

// ---------------------------------------------------------------- ScaleRating (1–10)

@Composable
fun JMScaleRating(
    value: Int?,
    onChange: (Int) -> Unit,
    modifier: Modifier = Modifier,
    range: IntRange = 1..10,
    lowLabel: String? = null,
    highLabel: String? = null,
) {
    Column(modifier, verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
            range.forEach { n ->
                val on = value == n
                Box(
                    Modifier
                        .weight(1f)
                        .height(44.dp)
                        .clip(JMShapes.sm)
                        .background(if (on) JMColors.ink else JMColors.surfaceCard)
                        .border(JMBorder.width, if (on) JMColors.ink else JMColors.borderDefault, JMShapes.sm)
                        .clickableNoRipple(true) { onChange(n) },
                    contentAlignment = Alignment.Center,
                ) {
                    Text("$n", color = if (on) JMColors.onPrimary else JMColors.textSecondary,
                        fontFamily = JMFontFamily.Sans, fontWeight = FontWeight.Bold, fontSize = 15.sp)
                }
            }
        }
        if (lowLabel != null || highLabel != null) {
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                Text(lowLabel ?: "", color = JMColors.textTertiary, fontSize = 12.sp)
                Text(highLabel ?: "", color = JMColors.textTertiary, fontSize = 12.sp)
            }
        }
    }
}

// ---------------------------------------------------------------- LockedAvatar (modesty)

@Composable
fun JMAvatar(
    initials: String = "",
    locked: Boolean = false,
    size: Dp = 48.dp,
    tint: Color = JMPalette.Pink100,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier.size(size).clip(CircleShape).background(if (locked) JMPalette.Ink200 else tint),
        contentAlignment = Alignment.Center,
    ) {
        if (locked) {
            // Replace with a Solar lock vector drawable; text shown as a placeholder.
            Text("\uD83D\uDD12".let { "" }, color = JMColors.textTertiary) // intentionally blank; use Icon(painterResource(R.drawable.ic_lock))
            Text("•", color = JMColors.textTertiary, fontSize = (size.value * 0.4).sp)
        } else {
            Text(initials, color = JMPalette.Pink700, fontFamily = JMFontFamily.Sans,
                fontWeight = FontWeight.Bold, fontSize = (size.value * 0.38).sp)
        }
    }
}

// ---------------------------------------------------------------- helpers

@Composable
private fun Modifier.clickableNoRipple(enabled: Boolean, onClick: () -> Unit): Modifier {
    val interaction = remember { androidx.compose.foundation.interaction.MutableInteractionSource() }
    return this.then(
        androidx.compose.foundation.clickable(
            interactionSource = interaction, indication = null, enabled = enabled, onClick = onClick,
        )
    )
}

File: native/ios/JustMarriage/Design/Components.swift
// Components.swift — Just Marriage SwiftUI component recipes.
// These mirror the web design system's contracts (see ../components.md). Copy what you
// need; they depend only on JMColor / JMTheme / JMFont.
import SwiftUI

// MARK: - Button

public enum JMButtonVariant { case primary, secondary, ink, outline, ghost }
public enum JMButtonSize { case sm, md, lg }

public struct JMButton: View {
    let title: String
    var variant: JMButtonVariant = .primary
    var size: JMButtonSize = .md
    var pill: Bool = false
    var fullWidth: Bool = false
    var systemIcon: String? = nil
    let action: () -> Void

    public init(_ title: String, variant: JMButtonVariant = .primary, size: JMButtonSize = .md,
                pill: Bool = false, fullWidth: Bool = false, systemIcon: String? = nil,
                action: @escaping () -> Void) {
        self.title = title; self.variant = variant; self.size = size
        self.pill = pill; self.fullWidth = fullWidth; self.systemIcon = systemIcon; self.action = action
    }

    private var height: CGFloat { switch size { case .sm: 36; case .md: 46; case .lg: 56 } }
    private var fontSize: CGFloat { switch size { case .sm: 13; case .md: 15; case .lg: 16 } }
    private var bg: Color {
        switch variant { case .primary: JMColor.primary; case .secondary: JMColor.secondary
        case .ink: JMColor.ink900; case .outline, .ghost: .clear }
    }
    private var fg: Color {
        switch variant { case .primary, .ink: JMColor.white; case .secondary: JMColor.onSecondary
        case .outline, .ghost: JMColor.ink900 }
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: 8) {
                if let icon = systemIcon { Image(systemName: icon) }
                Text(title).font(JMFont.sans(fontSize, .bold))
                    .lineLimit(1)
                    .minimumScaleFactor(0.86)
            }
            .frame(maxWidth: fullWidth ? .infinity : nil)
            .frame(height: height)
            .padding(.horizontal, size == .lg ? 30 : 22)
            .background(bg)
            .foregroundColor(fg)
            .overlay(
                RoundedRectangle(cornerRadius: pill ? JMRadius.pill : JMRadius.md, style: .continuous)
                    .strokeBorder(variant == .outline ? JMColor.ink900 : .clear, lineWidth: JMBorder.widthBold)
            )
            .clipShape(RoundedRectangle(cornerRadius: pill ? JMRadius.pill : JMRadius.md, style: .continuous))
        }
        .buttonStyle(JMPressStyle())
    }
}

/// Springy press scale used across interactive elements.
public struct JMPressStyle: ButtonStyle {
    public init() {}
    public func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.96 : 1)
            .animation(JMMotion.spring, value: configuration.isPressed)
    }
}

// MARK: - Badge (incl. tilted "100%" sticker tag)

public enum JMBadgeTone { case ink, pink, cyan, yellow, success, warning, error }

public struct JMBadge: View {
    let text: String
    var tone: JMBadgeTone = .ink
    var soft: Bool = false
    var tilt: Bool = false
    var uppercased: Bool = false

    public init(_ text: String, tone: JMBadgeTone = .ink, soft: Bool = false,
                tilt: Bool = false, uppercased: Bool = false) {
        self.text = text; self.tone = tone; self.soft = soft; self.tilt = tilt; self.uppercased = uppercased
    }

    private var pair: (bg: Color, fg: Color) {
        switch tone {
        case .ink:     return soft ? (JMColor.ink100, JMColor.ink900)   : (JMColor.ink900, JMColor.white)
        case .pink:    return soft ? (JMColor.pink50, JMColor.pink700)  : (JMColor.pink500, JMColor.white)
        case .cyan:    return soft ? (JMColor.cyan50, JMColor.cyan700)  : (JMColor.cyanBright, JMColor.ink900)
        case .yellow:  return (JMColor.yellow400, JMColor.ink900)
        case .success: return soft ? (JMColor.green50, JMColor.green500): (JMColor.green500, JMColor.white)
        case .warning: return soft ? (JMColor.amber50, JMColor.amber500): (JMColor.amber500, JMColor.ink900)
        case .error:   return soft ? (JMColor.red50, JMColor.red500)    : (JMColor.red500, JMColor.white)
        }
    }

    public var body: some View {
        Text(uppercased ? text.uppercased() : text)
            .font(JMFont.sans(13, .extrabold))
            .tracking(uppercased ? 0.5 : 0.1)
            .padding(.vertical, 4).padding(.horizontal, 11)
            .background(pair.bg)
            .foregroundColor(pair.fg)
            .clipShape(Capsule())
            .rotationEffect(.degrees(tilt ? -4 : 0))
    }
}

// MARK: - Card

public enum JMCardVariant { case plain, hard, tinted }

public struct JMCard<Content: View>: View {
    var variant: JMCardVariant = .plain
    var tint: Color = JMColor.cyan50
    var padding: CGFloat = JMSpace.x5
    @ViewBuilder let content: () -> Content

    public init(variant: JMCardVariant = .plain, tint: Color = JMColor.cyan50,
                padding: CGFloat = JMSpace.x5, @ViewBuilder content: @escaping () -> Content) {
        self.variant = variant; self.tint = tint; self.padding = padding; self.content = content
    }

    public var body: some View {
        let shape = RoundedRectangle(cornerRadius: JMRadius.lg, style: .continuous)
        return VStack(alignment: .leading, spacing: JMSpace.x3) {
            content()
        }
            .padding(padding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(variant == .tinted ? tint : JMColor.surfaceCard)
            .clipShape(shape)
            .overlay(
                shape.strokeBorder(
                    variant == .hard ? JMColor.ink900 : (variant == .plain ? JMColor.borderSubtle : .clear),
                    lineWidth: variant == .hard ? JMBorder.widthBold : JMBorder.width)
            )
            .background(
                variant == .hard
                    ? shape.fill(JMColor.ink900).offset(x: 4, y: 4)
                    : nil
            )
            .modifier(SoftShadow(on: variant == .plain))
    }
}

private struct SoftShadow: ViewModifier {
    let on: Bool
    func body(content: Content) -> some View {
        on ? AnyView(content.jmShadow(.sm)) : AnyView(content)
    }
}

// MARK: - ProgressRing (match-score dial)

public struct JMProgressRing: View {
    var value: Double            // 0...100
    var size: CGFloat = 120
    var thickness: CGFloat = 12
    var sublabel: String? = nil

    public init(value: Double, size: CGFloat = 120, thickness: CGFloat = 12, sublabel: String? = nil) {
        self.value = value; self.size = size; self.thickness = thickness; self.sublabel = sublabel
    }

    public var body: some View {
        ZStack {
            Circle().stroke(JMColor.ink100, lineWidth: thickness)
            Circle()
                .trim(from: 0, to: CGFloat(min(100, max(0, value)) / 100))
                .stroke(
                    AngularGradient(colors: [JMColor.pink500, JMColor.cyanBright],
                                    center: .center),
                    style: StrokeStyle(lineWidth: thickness, lineCap: .round))
                .rotationEffect(.degrees(-90))
                .animation(JMMotion.easeOut, value: value)
            VStack(spacing: 2) {
                Text("\(Int(value))%").font(JMFont.display(size * 0.3))
                    .foregroundColor(JMColor.ink900)
                if let s = sublabel {
                    Text(s.uppercased()).font(JMFont.sans(size * 0.1, .bold))
                        .tracking(1).foregroundColor(JMColor.textTertiary)
                }
            }
        }
        .frame(width: size, height: size)
    }
}

// MARK: - VoiceBars (counselor visualizer)

public struct JMVoiceBars: View {
    var active: Bool = true
    var barCount: Int = 7
    var tint: Color = JMColor.pink500
    var height: CGFloat = 80
    private static let idleBarScales: [CGFloat] = [0.32, 0.44, 0.36, 0.52, 0.36, 0.44, 0.32]
    @State private var phase = false

    public init(active: Bool = true, barCount: Int = 7, tint: Color = JMColor.pink500, height: CGFloat = 80) {
        self.active = active; self.barCount = barCount; self.tint = tint; self.height = height
    }

    public var body: some View {
        HStack(spacing: 6) {
            ForEach(0..<barCount, id: \.self) { i in
                Capsule().fill(tint)
                    .frame(width: 8, height: active ? height * barScale(i) : height * idleBarScale(i))
                    .opacity(active ? 1 : 0.82)
                    .animation(
                        active
                        ? .easeInOut(duration: 0.9).repeatForever().delay(Double(i % 4) * 0.12)
                        : .default,
                        value: phase)
            }
        }
        .frame(height: height)
        .onAppear { phase.toggle() }
    }
    private func barScale(_ i: Int) -> CGFloat { phase ? (i % 2 == 0 ? 1.0 : 0.5) : (i % 2 == 0 ? 0.3 : 0.9) }
    private func idleBarScale(_ i: Int) -> CGFloat {
        Self.idleBarScales[i % Self.idleBarScales.count]
    }
}

// MARK: - ScaleRating (1–10 importance)

public struct JMScaleRating: View {
    @Binding var value: Int?
    var range: ClosedRange<Int> = 1...10
    var lowLabel: String? = nil
    var highLabel: String? = nil

    public init(value: Binding<Int?>, range: ClosedRange<Int> = 1...10,
                lowLabel: String? = nil, highLabel: String? = nil) {
        self._value = value; self.range = range; self.lowLabel = lowLabel; self.highLabel = highLabel
    }

    public var body: some View {
        VStack(spacing: 10) {
            HStack(spacing: 6) {
                ForEach(Array(range), id: \.self) { n in
                    let on = value == n
                    Button { value = n } label: {
                        Text("\(n)").font(JMFont.sans(15, .bold))
                            .frame(maxWidth: .infinity, minHeight: 44)
                            .background(on ? JMColor.ink900 : JMColor.white)
                            .foregroundColor(on ? JMColor.white : JMColor.ink700)
                            .overlay(RoundedRectangle(cornerRadius: JMRadius.sm)
                                .strokeBorder(on ? JMColor.ink900 : JMColor.borderDefault, lineWidth: JMBorder.width))
                            .clipShape(RoundedRectangle(cornerRadius: JMRadius.sm))
                            .offset(y: on ? -2 : 0)
                    }.buttonStyle(.plain)
                }
            }
            if lowLabel != nil || highLabel != nil {
                HStack {
                    Text(lowLabel ?? "").font(JMFont.sans(12)).foregroundColor(JMColor.textTertiary)
                    Spacer()
                    Text(highLabel ?? "").font(JMFont.sans(12)).foregroundColor(JMColor.textTertiary)
                }
            }
        }
    }
}

// MARK: - LockedAvatar (modesty model)

public struct JMAvatar: View {
    var initials: String = ""
    var locked: Bool = false
    var size: CGFloat = 48
    var tint: Color = JMColor.pink100

    public init(initials: String = "", locked: Bool = false, size: CGFloat = 48, tint: Color = JMColor.pink100) {
        self.initials = initials; self.locked = locked; self.size = size; self.tint = tint
    }

    public var body: some View {
        ZStack {
            Circle().fill(locked ? JMColor.ink200 : tint)
            if locked {
                Image(systemName: "lock.fill").font(.system(size: size * 0.4)).foregroundColor(JMColor.ink500)
            } else {
                Text(initials).font(JMFont.sans(size * 0.38, .bold)).foregroundColor(JMColor.pink700)
            }
        }
        .frame(width: size, height: size)
    }
}
